### ISSUE TITLE
시스템18-2 성능실험

### ISSUE THREAD

**Issue by bart2001** [nlp-research/biaffineparser#66]
시스템 특징
- Pre-trained 단어/태그 embedding (초기 200-dim, trainble=False, embeddings/words.pos.original.vec)
- Random 음절 embedding (초기 50-dim, trainble=True)
- Random 태그 embedding (초기 100-dim, trainble=True)
- Bi-LSTM at embedding_layer
- num_lstm_layer=3

* (해결) 학습데이터에 ;이 포함된 문장에 대해서 코멘트 처리가 되어서 nan으로 해석된 사례

**Comment by bart2001** [comment]
fixed in parser/model.py

```python
   with tf.variable_scope('char'):
                trainable = False if self.hparams.char_embed_file else True
                _char_embedding = tf.Variable(
                    self.char_embedding, trainable=trainable,
                    name="_char_embedding", dtype=tf.float32)
                char_embedding = tf.nn.embedding_lookup(
                    _char_embedding, self.char_ids, name="char_embedding")
                shape = tf.shape(char_embedding)
                '''BEFORE
                char_embedding = tf.reshape(
                    char_embedding, [-1, shape[2], self.hparams.char_embed_size])
                char_embedding = bilstm_layer(char_embedding, sequence_length, int(
                    self.hparams.char_embed_size / 2))
                char_embedding = tf.reshape(
                    char_embedding, [-1, shape[1], self.hparams.char_embed_size])
                '''
                #AFTER
                char_embedding = tf.reshape(
                    char_embedding, [-1, shape[2], self.hparams.char_embed_size])
                char_embedding = bilstm_layer(char_embedding, sequence_length, 100) #100
                char_embedding = tf.reshape(
                    char_embedding, [-1, shape[1], 200]) #200
                if self.embed_dropout > 0.0:
                    keep_prob = 1.0 - self.embed_dropout
                    char_embedding = tf.nn.dropout(
                        char_embedding, keep_prob)
```

**Comment by bart2001** [comment]
모델18과 비교하여
multilayer_perceptron에 output_layer가 적용되지 않았던 문제는 해결된 상태 (성능 향상 예상)

in parser/model.py
```python
def multilayer_perceptron(x, weights, biases, dropout):
    layer_1 = tf.add(tf.matmul(x, weights['w1']), biases['b1'])
    layer_1 = tf.nn.relu(layer_1)
    out_layer = tf.add(tf.matmul(layer_1, weights['w2']), biases['b2'])
    if dropout > 0.0:
        keep_prob = 1.0 - dropout
        #layer_1 = tf.nn.dropout(layer_1, keep_prob)
        out_layer = tf.nn.dropout(out_layer, keep_prob)
    return out_layer #NOT layer1
```

**Comment by bart2001** [comment]
hparams
```python
{
    "num_lstm_units": 400,
    "num_lstm_layers": 3,
    "arc_mlp_units": 500,
    "label_mlp_units": 100,
    "word_embed_size": 200,
    "char_embed_size": 50,
    "pos_embed_size": 100,
    "optimizer": "adam",
    "learning_rate": 0.001,
    "decay_factor": 0.9,
    "num_train_epochs": 50,
    "train_filename": "data/sejong.nan.fixed.train.csv",
    "dev_filename": "data/sejong.nan.fixed.test.csv",
    "out_dir": "parser/model_18_2",
    "word_vocab_name": "word.pkl",
    "char_vocab_name": "char.pkl",
    "pos_vocab_name": "pos.pkl",
    "rel_vocab_name": "rel.pkl",
    "head_vocab_name": "head.pkl",
    "word_embed_file": "embeddings/words.pos.original.vec",
    "char_embed_file": "",
    "pos_embed_file": "",
    "word_embed_matrix_file": "embeddings/word_embed_matrix.txt",
    "char_embed_matrix_file": "embeddings/char_embed_matrix.txt",
    "pos_embed_matrix_file": "embeddings/pos_embed_matrix.txt",
    "embed_dropout": 0.33,
    "lstm_dropout": 0.33,
    "mlp_dropout": 0.33,
    "batch_size": 128,
    "device": "gpu",
    "debug": false
}
```

최고성능
```python
epoch: 4 - loss: 0.3515779972076416 - uas: 0.9060402512550354 - las: 0.890380322933197 - eval_loss: 0.36788931488990784 - eval_uas: 0.9164566397666931 - eval_las: 0.8939207196235657
# new best UAS!
```
